[Core] Neighbor MPC Assigner: A process and utility for assigning MPCs to neighbouring nodes

SADPR opened the pull request:
### DESCRIPTION:

**Utility:**

This pull request adds a new class that provides a method to search for neighboring nodes of a given nodes structure within a specified radius and assign a multipoint constraint (MPC). The MPC weights are calculated based on a spatial function that employs radial basis functions.

**Process:**

This pull request enhances the functionality of the master-slave MPC algorithm by facilitating the discovery of neighboring nodes in a master model part within a designated radius for each node in the slave model part. The process involves the following steps:

1. For each node in the slave model part, the utility searches for neighboring nodes in the master model part within a specified radius.
![image](https://user-images.githubusercontent.com/61457043/236417112-8c5b1c89-9e43-4467-bba0-241ad994b17a.png)
![image](https://user-images.githubusercontent.com/61457043/236417146-46b90785-4b0a-48c0-8949-f9ba2abed309.png)
![image](https://user-images.githubusercontent.com/61457043/236417228-ba3a7a9c-519f-452c-b2db-69fd1226a06e.png)

3. The utility then establishes an MPC between the slave node and its neighboring master nodes, and calculates the MPC weights based on a spatial function that employs radial basis functions.
![image](https://user-images.githubusercontent.com/61457043/236417284-bcaa49ba-340e-45de-9fc8-695766648072.png)
![image](https://user-images.githubusercontent.com/61457043/236417336-6e117e24-5a91-4903-ab4c-d302b29ae576.png)
4. The process repeats for all nodes in the slave model part.

**Added files:**

- `assign_mpcs_to_neighbours_utility.cpp`: This file contains all the functions to look for neighboring nodes and apply multipoint constraints (MPCs) using radial basis functions.
- `assign_mpcs_to_neighbours_utility.h`: This header file defines the functions and data structures used in assign_mpcs_to_neighbours_utility.cpp`.
- `assign_mpcs_to_neighbours_process.py`: This Python script calls the functions in `assign_mpcs_to_neighbours_utility.cpp` in the appropriate order to assign MPCs to neighboring nodes.
- `assign_mpcs_to_neighbours_process.mdpa`: This is an MDPA file used for testing the MPC algorithm.

**Modified files:**

- `add_geometrical_utilities_to_python.cpp`: This file was modified to export the functions in `assign_mpcs_to_neighbours_utility.cpp` to Python, so that they can be called from `assign_mpcs_to_neighbours_process.py`.
- `test_processes.py`: This file was modified to include a small test to validate the functionality of `assign_mpcs_to_neighbours_process.py`.

SADPR committed:
Working Assign MPCs to niehgbors process and utility 

It is currently done in serial.

SADPR committed:
Cleaned. Not working in parallel.

It worked when using vector variable, e.g. VELOCITY

SADPR committed:
Working in parallel.

SADPR committed:
Added test.

loumalouomega committed:
Minor changes

loumalouomega commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,179 @@
+    def RemoveConstraints(self):
+        #Remove master-slave constraints
+        for constraint in self.computing_model_part.MasterSlaveConstraints:
+            constraint.Set(KM.TO_ERASE)
```
Use VariableUtils

loumalouomega commented on kratos/tests/auxiliar_files_for_python_unittest/mdpa_files/assign_mpcs_to_neighbours_process.mdpa:
```diff
@@ -0,0 +1,3308 @@
+Begin Properties 0
```
Please, can you consider a smaller mesh?, you can put this one in the Example repository

loumalouomega commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,211 @@
+
+    ModelPart::MasterSlaveConstraintType const& r_clone_constraint = KratosComponents<MasterSlaveConstraint>::Get("LinearMasterSlaveConstraint");
+    
+    #pragma omp parallel
```
Please ParallelUtilities

loumalouomega commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,211 @@
+DofPointerVectorType& rCloud_of_dofs,
+array_1d<double,3>& rSlave_coordinates
+){
+KRATOS_TRY;
```
Tabs missing (spaces)

loumalouomega commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,211 @@
+Matrix& rCloud_of_nodes_coordinates
+)
+{
+KRATOS_TRY;
```
Tabs missing (spaces)

loumalouomega commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,211 @@
+)
+{
+//TODO: Do it in parallel
+KRATOS_TRY;
```
Tabs missing (spaces)

loumalouomega reviewed (changes requested):
Some comments

loumalouomega committed:
Missing spaces

SADPR committed:
Changed to VariabelUtils.

SADPR committed:
Merge remote-tracking branch 'origin/Kratos_Assign_MPCs' into Kratos_Assign_MPCs

SADPR committed:
Changing to Parallel Utilities. It will fail. This is not ready.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,282 @@
+        Matrix r_cloud_of_nodes_coordinates;
+        array_1d<double,3> r_slave_coordinates;
+        GetDofsAndCoordinatesForNode(pNode, rVariable, r_slave_dof, r_slave_coordinates);
+        KRATOS_WATCH(i)
```
```suggestion
```
Debugging leftover.

SADPR committed:
Not working properly. Trying to incorporate KRATOS_CRITICAL_SECTION.

SADPR committed:
Updated header file.

SADPR committed:
Added Parallel Utilities with std::mutex and std::lock_guard. Too slow.

Way slower than OMP and I couldn't manage to do the omp critical for adding the MPCs.

SADPR committed:
Updated the python file to use the ParallelUtilities function.

SADPR committed:
Minor change for importing to python (Name update).

SADPR committed:
Working with Parallel Utilities.

SADPR committed:
Code cleaning.

SADPR committed:
Erased auxiliary function.

SADPR committed:
Updating SearchNodesInRadiusForNodesParallelUtilities.

SADPR committed:
Erasing OMP from every function.

SADPR committed:
Working for 3d variables as well.

SADPR committed:
Adapted the last function to be adapted.

SADPR committed:
Erased previous mdpa (too large).

SADPR committed:
Updated mdpa, very coarse.

SADPR committed:
Updated test results to fit new mdpa.

SADPR committed:
Erasing this DoF.

SADPR committed:
Last minor change. Ready to merge (if possible).

SADPR committed:
This is the last commit x3.

loumalouomega commented on kratos/tests/auxiliar_files_for_python_unittest/mdpa_files/test_assign_mpcs_to_neighbours_process.mdpa:
```diff
@@ -0,0 +1,220 @@
+       48          0    17    16    20 
+End Elements
+
+Begin SubModelPart GENERIC_Domain // Group Domain // Subtree GENERIC
```
If this submodelpart is not used you can simply removed to make the mdpa even smaller. I understand you only need master/slave ones

SADPR replied:
Since some applications require the MPCs to be applied to an specific model part (e.g. FluidDynamics -> fluid_computing_model_part). The process needs the name of this model part. In this case, GENERIC_Domain plays this role.

loumalouomega commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,379 @@
+}
+
+
+
```
```suggestion

```

loumalouomega reviewed (commented):
Minor comments

SADPR committed:
Update kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

Co-authored-by: Vicente Mataix Ferrándiz <vmataix@altair.com>

rubenzorrilla reviewed (commented):
I've some comments about this. Will post them soon.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+    return AssignMPCsToNeighboursProcess(model, settings["Parameters"])
+
+## All the processes python should be derived from "Process"
+class AssignMPCsToNeighboursProcess(KM.Process):
```
I'd call this `AssignMasterSlaveConstraintsToNeighboursProcess` to follow the naming convention of the rest of the Kratos (see the `ModelPart` for instance).

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+        KM.Process.__init__(self) # calling the baseclass constructor
+
+        default_settings = KM.Parameters("""{
+            "computing_model_part_name": "",
```
`model_part_name` to be consistent with the rest of processes. Also note that not all the solvers in Kratos require a computing model part.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+            "variable_names": [],
+            "search_radius": 1.0,
+            "minimum_number_of_neighbouring_nodes": 3,
+            "assign_mpcs_every_time_step": false
```
I think that `reform_constraints at_each_step` is more Kratos-like (see the `reform_dofs_at_each_step` in the solvers).

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+        # Add missing settings that the user did not provide but that
+        # are necessary for this process
+        settings.ValidateAndAssignDefaults(default_settings)
+        # Alternative:
+        # settings.RecursivelyValidateAndAssignDefaults(default_settings)
```
Remove commented code.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+            if not KM.KratosGlobals.GetVariableType(var_name):
+                err_msg = "\'{}\' variable in \'variable_names\' is not double type. Please check provide double type variables (e.g. [\"DISPLACEMENT_X\",\"DISPLACEMENT_Y\"]).".format(var_name)
+            self.variables_list.append(KM.KratosGlobals.GetVariable(var_name))
+        debug = True
```
What is this?

SADPR replied:
Debugging leftover.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        ## Assign master and slave model parts
+        self.computing_model_part = self.model.GetModelPart(self.computing_model_part_name)
+        self.slave_model_part = self.model.GetModelPart(self.slave_model_part_name)
+        self.master_model_part = self.model.GetModelPart(self.master_model_part_name)
```
You have member variables for both model and model parts. Either one or the other. In this case it might be easier just to move these three lines to the constructor and avoid saving the model.

If you're doing this because you cannot assume that at construction time of this process the model will be filled, please put a comment explaining so, as this is not common for processes.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+            for variable in self.variables_list:
+                self.assign_mpcs_utility.AssignMPCsToNodes(self.slave_model_part.Nodes,self.search_radius,self.computing_model_part, variable, self.minimum_number_of_neighbouring_nodes)
+
+    def Check(self):
+        """ This method verifies that the input is correct
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        pass
+
+    def ExecuteBeforeSolutionLoop(self):
+        """ This method is executed just before the solution-loop
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        pass
```
Remove if not implemented.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+            for variable in self.variables_list:
+                self.assign_mpcs_utility.AssignMPCsToNodes(self.slave_model_part.Nodes,self.search_radius,self.computing_model_part, variable, self.minimum_number_of_neighbouring_nodes)
+
+    def ExecuteBeforeOutputStep(self):
+        """ This method is executed before writing the output (if output
+        is being written in this step)
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        pass
+
+    def ExecuteAfterOutputStep(self):
+        """ This method is executed after writing the output (if output
+        is being written in this step)
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        pass
```
Remove if not implemented.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+        KM.VariableUtils().SetFlag(KM.TO_ERASE, True, self.computing_model_part.MasterSlaveConstraints)
+        self.computing_model_part.RemoveMasterSlaveConstraintsFromAllLevels(KM.TO_ERASE)
+
+    def ExecuteFinalize(self):
+        """ This method is executed after the computations, at the end of the solution-loop
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        pass
```
Remove if not implemented.

rubenzorrilla commented on kratos/python_scripts/assign_mpcs_to_neighbours_process.py:
```diff
@@ -0,0 +1,178 @@
+        if self.assign_mpcs_every_time_step:
+            self.RemoveConstraints()
+        
+    def RemoveConstraints(self):
```
Make this "private" by mangling it. Otherwise you'd be extending the process API.

rubenzorrilla commented on kratos/python/add_geometrical_utilities_to_python.cpp:
```diff
@@ -432,7 +432,18 @@ void AddGeometricalUtilitiesToPython(pybind11::module &m)
         .def_static("CalculateShapeFunctions", [](const Matrix& rPoints, const array_1d<double,3>& rX, const double h, Vector& rN, DenseQRPointerType pDenseQR){
             return RBFShapeFunctionsUtility::CalculateShapeFunctions(rPoints, rX, h, rN, pDenseQR);})
         ;
+    
```
See my comment about the Kratos naming (change `MPC` to `MasterSlaveConstraint`).

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+      using NodesContainerType = ModelPart::NodesContainerType;
+
+      //Bin
+      using NodeBinsType = BinsDynamic<3, NodeType, NodesContainerType::ContainerType>;
```
Shouldn't the first template argument be 2 for the 2D case?

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+      ///@name Life Cycle
+      ///@{
+
+      /// Default constructor.
```
This is not the default constructor. Indeed, I'd delete the default constructor (it's being automatically added by the compiler).

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+      Kratos::unique_ptr<NodeBinsType> mpBins;
+      int mMaxNumberOfNodes;
+
+      /// Assignment operator.
+      AssignMPCsToNeighboursUtility& operator=(AssignMPCsToNeighboursUtility const& rOther)
+      {
+          return *this;
+      }
```
I'd remove the assignment operator.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+    private:
+      ///@name Member Variables
+      ///@{
+      Kratos::unique_ptr<NodeBinsType> mpBins;
```
NodeBinsType::UniquePointer

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+      void GetDofsAndCoordinatesForNode(
+        NodeType::Pointer pNode,
+        const Variable<array_1d<double, 3>>& rVariable,
+        DofPointerVectorType& rCloud_of_dofs_x,
+        DofPointerVectorType& rCloud_of_dofs_y,
+        DofPointerVectorType& rCloud_of_dofs_z,
```
Please follow the style guide. CamelCase for method arguments.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+       * @param RotationMatrix 3x3 rotation matrix to be applied to the nodes' positions.
+       **/
+      void AssignRotationToNodes(
+          NodesContainerType pNodes,
```
This should be a reference. Indeed it is strange to me that this works if you pass it by copy...

SADPR replied:
Not used anymore.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.h:
```diff
@@ -0,0 +1,252 @@
+
+      // Search for the neighbouring nodes (in rStructureNodes) of each rNode on a given array of rNodes
+      void SearchNodesInRadiusForNodes(
+          NodesContainerType const& rNodes,
```
Though it is valid, in Kratos we normally put the `const` in front.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+// Search for the neighbouring nodes (in rStructureNodes) of each rNode on a given array of rNodes
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNodes(
+    NodesContainerType const& rNodes,
```
Normally in Kratos we put the `const` in front.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNodes(
+    NodesContainerType const& rNodes,
+    const double Radius,
+    VectorResultNodesContainerType& rResults,
+    const double MinNumOfNeighNodes)
```
It is a good programming practice to put the results arguments after the input (const ones).

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    const double MinNumOfNeighNodes)
+{
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(rNodes.GetContainer());
```
`auto& r_nodes_array`

rubenzorrilla replied:
And why you need the `const_cast`?

SADPR replied:
Resolved, changed to:
```python
    KRATOS_TRY;
    auto& r_nodes_array = rNodes.GetContainer();

    rResults.resize(r_nodes_array.size());

    // Declare a counter variable outside the loop as std::atomic<int>
    std::atomic<int> i(0);

    block_for_each(r_nodes_array, [&](const Node<3>::Pointer& pNode)
    {
```

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    block_for_each(nodes_array, [&](Node<3>::Pointer& pNode)
+    {
+        double localRadius = Radius;
```
Please follow the style guide (`local_radius`).

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(rNodes.GetContainer());
+
+    rResults.resize(nodes_array.size());
```
This means you're allocating much memory than needed. I'd do a `shrink_to_fit` at the end of this function to avoid so.

SADPR replied:
I am allocating what I need. Nonetheless, I could updated it to this:
```cpp
    // Resize rResults vector
    rResults.resize(r_nodes_array.size());

    // Declare a counter variable outside the loop as std::atomic<int>
    std::atomic<int> i(0);

    block_for_each(r_nodes_array, [&](const Node<3>::Pointer& pNode)
    {
        double local_radius = Radius;

        IndexType it = i.fetch_add(1); // Atomically increment the counter and get the previous value

        while (rResults[it].size()<MinNumOfNeighNodes){
            rResults[it].clear();
            SearchNodesInRadiusForNode(pNode, local_radius, rResults[it]);
            local_radius += Radius;
        }
    });

    // Optimize memory usage
    rResults.shrink_to_fit();

rubenzorrilla replied:
No. You are allocating much more than what you need since you're resizing the vector to the number of nodes, which would only be the case if all the nodes are found. Indeed, the proper usage would be
```c++
rResults.reserve(r_nodes_array.size());

// ... your stuff in here ...

rResults.shrink_to_fit();
```

SADPR replied:
I am changing the names of the functions to make them more self-explanatory. In this function, we retrieve the degrees of freedom (DOFs) and coordinates of the cloud of nodes. Therefore, we already know the size.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+//Search for the neighbouring nodes (in rStructureNodes) of the given rNode
+void AssignMPCsToNeighboursUtility::SearchNodesInRadiusForNode(
+    NodeType::Pointer pNode,
```
I'd pass a constant reference instead. Note that the `SearchInRadius` supports it.

SADPR replied:
Do you find it necesarry, since this is a constant pointer. Isn't it okay like this?

rubenzorrilla replied:
I don't think it is a constant pointer but nevermind, leave it like this.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    if (ComponentIndex < 3) {
+        std::string component_variable_name = rVectorVariable.Name() + component_suffixes[ComponentIndex];
+        return KratosComponents<Variable<double>>::Get(component_variable_name);
```
This is extremely expensive and being done, if I'm not wrong, inside a loop. An alternative approach must be used.

SADPR replied:
After a long afternoon, I managed to do it. Now it only acces the Kratos components once and has a pointer in the loop.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    )
+{
+    KRATOS_TRY;
+    rCloud_of_dofs.push_back(pNode->pGetDof(rVariable));
+    rSlave_coordinates = pNode->Coordinates();
```
I don't see the point on doing a function of two lines. IMO it only makes more complex browsing the code with no advantage at all.

SADPR replied:
It is to preserve similarity. Is it really necessary to undo this function? I know it is not typical to have a two line function, but for me, it preserves similarity.

rubenzorrilla replied:
I don't get your point on "similarity"... I'd remove it but if you really require it leave it. Up to you.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    KRATOS_TRY;
+
+    // Check if the node has the required DOFs
+    if (pNode->HasDofFor(GetComponentVariable(rVariable, 0)) &&
+        pNode->HasDofFor(GetComponentVariable(rVariable, 1)) &&
+        pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
```
What would happen in the 2D case in which the third component DOF might not be present?

SADPR replied:
Is this really something that happens? 
The user can also specify ["DISPLACEMENT_X","DISPLACEMENT_Y"].

SADPR replied:
I thought the 2D also 3 components. But solves only for X and Y components.

rubenzorrilla replied:
Depends on the implementation so you cannot take it as granted.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    DofPointerVectorType& rCloud_of_dofs_x,
+    DofPointerVectorType& rCloud_of_dofs_y,
+    DofPointerVectorType& rCloud_of_dofs_z,
+    array_1d<double,3>& rSlave_coordinates
```
Please follow the style guide for function arguments.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+
+void AssignMPCsToNeighboursUtility::GetDofsAndCoordinatesForNodes(
+    ResultNodesContainerType nodes_array,
```
Why passing by copy?

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+)
+{
+    KRATOS_TRY;
+    rCloud_of_dofs_x.resize(nodes_array.size());
+    rCloud_of_dofs_y.resize(nodes_array.size());
+    rCloud_of_dofs_z.resize(nodes_array.size());
+    rCloud_of_nodes_coordinates.resize(nodes_array.size(), 3);
```
I'd try to shrink these containers afterwards.

rubenzorrilla replied:
(same `reseve` and `shrink_to_fit` comment).

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+        // Check if the node has the required DOFs
+        if (pNode->HasDofFor(GetComponentVariable(rVariable, 0)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 1)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
```
Same comment about the 2D third component.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+{
+    //TODO: Do it in parallel
+    KRATOS_TRY;
+    NodesContainerType::ContainerType& nodes_array     = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
```
`auto& r_nodes_array`

rubenzorrilla replied:
Why the `const_cast`?

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    // Assign rotation to given nodes
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        auto coordinate_vector = nodes_array[i]->GetInitialPosition().Coordinates();
```
You are allocating/deallocating an unnecessary copy for each node.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    // Assign rotation to given nodes
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        auto coordinate_vector = nodes_array[i]->GetInitialPosition().Coordinates();
+        Vector rotated_position(coordinate_vector.size());
```
Can't this be an `array_1d<double,3>`?

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+        auto coordinate_vector = nodes_array[i]->GetInitialPosition().Coordinates();
+        Vector rotated_position(coordinate_vector.size());
+        noalias(rotated_position) = prod(RotationMatrix,coordinate_vector);
+        nodes_array[i]->X() = rotated_position[0];
+        nodes_array[i]->Y() = rotated_position[1];
+        nodes_array[i]->Z() = rotated_position[2];
```
`noalias(nodes_array[i]->Coordinates()) = rotated_position;` I think this should work.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignMPCsToNodes(
+    NodesContainerType pNodes,
+    double const Radius,
+    ModelPart& rComputingModelPart,
+    const Variable<array_1d<double, 3>>& rVariable,
+    double const MinNumOfNeighNodes
+    )
```
This is basically copy & paste of the above function but modifying the section in which the Master-Slave Constraints are created. I'd do the creation within a private auxiliary function and have a unique implementation of `AssignMPCsToNodes`.

SADPR replied:
I have a concern here, at first I though of the user passing a variable which could be either array_1d or double.
If array_1d, then you have to create the variable components lists, etc. If double, you can directly retrieve the varible with GetDof. Now I am thinking to always create a list of variables, e.g, if ["VELOCITY","PRESSURE], then create a list:
["DISPLACEMENT_X","DISPLACEMENT_Y","PRESSURE"] for 2D.
Or:
["DISPLACEMENT_X","DISPLACEMENT_Y","DISPLACEMENT_Z","PRESSURE"] for 3D.
We can discuss this in person, since it is taking a lot of time.

rubenzorrilla replied:
OK.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    // Add the constraints to the rComputingModelPart in a single call
+    rComputingModelPart.AddMasterSlaveConstraints(temp_constraints.begin(), temp_constraints.end());
+
+    KRATOS_INFO("AssignMPCsToNeighboursUtility") << "Build and Assign MPCs Time: " << build_and_assign_mpcs.ElapsedSeconds() << std::endl;
```
This may result in a too verbose output. If you want to output this, I'd introduce an echo level.

SADPR replied:
TODO.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    BuiltinTimer build_and_assign_mpcs;
+
+    int prev_num_mpcs = rComputingModelPart.NumberOfMasterSlaveConstraints();
```
In here you are assuming that your new MPCs are not interacting with previous ones. Though this might be a valid assumption, I'd throw a warning message if the previous number of MPCs is non-zero to inform about this.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    int prev_num_mpcs = rComputingModelPart.NumberOfMasterSlaveConstraints();
+
+    NodesContainerType::ContainerType& nodes_array = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
```
`auto& r_nodes_array`.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    NodesContainerType::ContainerType& nodes_array = const_cast<NodesContainerType::ContainerType&>(pNodes.GetContainer());
+
+    ModelPart::MasterSlaveConstraintType const& r_clone_constraint = KratosComponents<MasterSlaveConstraint>::Get("LinearMasterSlaveConstraint");
```
const auto& r_clone_constraint`.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+        }
+
+        // Get Dofs and Coordinates
+        DofPointerVectorType r_cloud_of_dofs, r_slave_dof;
+        Matrix r_cloud_of_nodes_coordinates;
+        array_1d<double, 3> r_slave_coordinates;
```
You're allocating/deallocating these for each node. Declare them as a TLS outside the loop.

SADPR replied:
Master-Slave constraints require rSlaveDofs to be a DofPointerVectorType. To do r_slave_dofs thread_local I need to resize and assign to Dof<double>::Pointer. Or:
rCloudOfDofs.resize(1); 
rCloudOfDofs[0] = pNode->pGetDof(rVariable);
Which I dont like. I will fix it.

rubenzorrilla commented on kratos/tests/test_processes.py:
```diff
@@ -2521,6 +2521,84 @@ def test_split_internal_interfaces_process_3D(self):
+        )
+
+        # Create and execute processes
+        list_of_processes = process_factory.KratosProcessFactory(current_model).ConstructListOfProcesses(settings["process_list"])
+        for process in list_of_processes:
+            process.ExecuteInitialize()
+            process.ExecuteInitializeSolutionStep()
```
You have a unique process. I don't see the point on doing it with a list. I'd directly call the constructor.

rubenzorrilla commented on kratos/tests/test_processes.py:
```diff
@@ -2521,6 +2521,84 @@ def test_split_internal_interfaces_process_3D(self):
+
+            for obtained_weights in obtained_constraints_weights:
+                if check_matching_weights(expected_weights, obtained_weights):
+                    for ew, ow in zip(expected_weights, obtained_weights):
+                        self.assertAlmostEqual(ew, ow, delta=1e-6)  # Assert the weights are equal within the specified tolerance
```
If I'm not wrong, there is an `assertAlmostEqualVector` or something of this sort.

SADPR replied:
assertVectorAlmostEqual

rubenzorrilla reviewed (commented):
Though this is a working implementation, I think it doesn't suffice the quality requirements for being in the `KratosCore`. I hope you take my comments into consideration in a new PR.

rubenzorrilla commented on kratos/utilities/assign_mpcs_to_neighbours_utility.cpp:
```diff
@@ -0,0 +1,376 @@
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignRotationToNodes(
```
This is not used anywhere in this process. Besides, it is a custom function for your target application. Long story short, it must be removed from here to be placed somewhere else.
